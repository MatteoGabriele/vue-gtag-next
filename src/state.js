import { ref, reactive, computed, toRefs } from "vue";

const state = reactive({
  property: null,
  isEnabled: true,
  globalObjectName: "gtag",
  globalDataLayerName: "dataLayer",
  domain: "https://www.googletagmanager.com",
  customResource: null,
  useDebugger: false,
});

export const routeState = reactive({
  template: null,
  appName: null,
  useScreenview: false,
  skipSamePath: true,
});

export const useState = () => toRefs(state);

export const useRouteState = () => toRefs(routeState);

export const isBootstrapped = ref(false);

export const isReady = ref(false);

export const isTrackRouterEnabled = ref(false);

export const defaultProperty = computed(() => {
  const { property } = useState();

  if (!property.value) {
    return;
  }

  if (Array.isArray(property.value)) {
    return property.value.find((p) => p.default === true) || property.value[0];
  }

  return property.value;
});

export const hasId = computed(() => {
  const { property } = useState();
  return property.value && property.value.id !== null;
});

export const allProperties = computed(() => {
  const { property } = useState();

  if (Array.isArray(property.value)) {
    return property.value;
  }

  return [property.value];
});

export const isTracking = computed(() => {
  const { isEnabled } = useState();
  const property = defaultProperty.value;

  return property && property.id && isEnabled.value;
});

export default state;
