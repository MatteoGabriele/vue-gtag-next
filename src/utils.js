export const loadScript = (source, domain) => {
  return new Promise((resolve, reject) => {
    const head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");

    script.async = true;
    script.src = source;
    script.charset = "utf-8";

    if (domain) {
      const link = document.createElement("link");

      link.href = domain;
      link.rel = "preconnect";

      head.appendChild(link);
    }

    head.appendChild(script);

    script.onload = resolve;
    script.onerror = reject;
  });
};

export const merge = (obj = {}, newObj = {}) => {
  Object.keys(newObj).forEach((key) => {
    obj[key] = newObj[key];
  });
};
