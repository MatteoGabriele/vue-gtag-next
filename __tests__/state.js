import state, { defaultProperty } from "@/state";
import { merge } from "@/utils";

const defaultState = { ...state };

describe("options", () => {
  beforeEach(() => {
    merge(state, defaultState);
  });

  afterEach(() => {
    jest.restoreAllMocks();
    jest.clearAllMocks();
  });

  it("should render default plugin state", () => {
    expect(state).toMatchSnapshot();
  });

  describe("defaultProperty", () => {
    it("should return the object", () => {
      merge(state, {
        property: {
          id: 1,
        },
      });

      expect(defaultProperty.value).toEqual({ id: 1 });
    });

    it("should return the only item in the array", () => {
      merge(state, {
        property: [
          {
            id: 1,
          },
        ],
      });

      expect(defaultProperty.value).toEqual({ id: 1 });
    });

    it("should return the only item in the array", () => {
      merge(state, {
        property: [
          {
            id: 1,
          },
          {
            id: 2,
            default: true,
          },
        ],
      });

      expect(defaultProperty.value).toEqual({ id: 2, default: true });
    });
  });
});
