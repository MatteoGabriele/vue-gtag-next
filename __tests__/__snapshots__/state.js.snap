// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options should render default plugin state 1`] = `
Object {
  "customResource": null,
  "domain": "https://www.googletagmanager.com",
  "globalDataLayerName": "dataLayer",
  "globalObjectName": "gtag",
  "isEnabled": true,
  "property": null,
  "useDebugger": false,
}
`;
